Apple pay - Best way forward to verify domain

ricardo opened the pull request:
Fixes #1376
Fixes #1442
Fixes #1330

This PR is complementary to #1465 in improving domain registration reliability.

The default method (rewrite rule) doesn't work when permalinks are set to Plain, so we try to copy the domain association file to the domain root as a fallback. This fallback may not work for lack of permissions in modifying the .well-known folder.

# Changes proposed in this Pull Request:

- Restore file copy method and adapt it as a fallback.
- Add tests to both verification methods.
- Trigger domain verification upon domain name change.
- Fix duplicate registration failure inbox notices:
  By deleting previously generated notes and creating a new failure note if necessary. I first tried managing soft deleted notes or unactioned notes, but the code became complex and I thought it would be better to simply delete all notes and then recreate one if verification fails - It's also better because it pushes the newest error note to the top.

# Testing instructions

## Test domain registration with fallback method

- Make sure you have live publishable and secret keys in settings.
- Go to [http://localhost:8082/wp-admin/options-permalink.php](http://localhost:8082/wp-admin/options-permalink.php) and set your permalink setting to `Plain`. This will break the rewrite rule method.
- Make sure you have a new unverified domain. You can use [Ngrok](https://ngrok.com): `ngrok http 8082`.
- Open any wp-admin page of your unverified domain. - A new domain registration request shall be triggered.
- Check any HTTPS product URL using Safari (Mac or iOS), with an active payment method in Apple Wallet. You should be able to see the Apple Pay button.

## Testing duplicate inbox note

For the #1376 issue, you'll want to make the domain registration fail. You can easily test this by trying to verify `localhost:8082`.

- Open your [localhost Stripe](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe) settings.
- Disable Payment Request Buttons, save, then enable and save again.
- You should see the following Inbox Notice in your WooCommerce Home: "Apple Pay domain verification needed".
- Try repeating the process and check if there aren't duplicate notices.
- Try validating your domain with Ngrok and check if the note disappears.

Note: If you keep a wp-admin tab open with different URLs, it can trigger the domain registration (by domain name change) with timed Ajax calls - So make sure you don't have wp-admin with different URLs open at the same time.

ricardo committed:
Add/update domain association file as a fallback

ricardo committed:
Fix duplicate inbox notes

ricardo committed:
Refactor domain verification failure notes

ricardo committed:
Add unit tests

ricardo committed:
Verify domain upon domain name change.

reykjalin commented on includes/admin/class-wc-stripe-inbox-notes.php:
```diff
@@ -57,28 +57,26 @@ public static function notify_on_apple_pay_domain_verification( $verification_co
+				$note    = WC_Admin_Notes::get_note( $note_id );
+				$note->delete();
+			}
+		} catch ( Exception $e ) {
+			return;
+		}
+
 		if ( $verification_complete ) {
 			if ( self::should_show_marketing_note() && ! wp_next_scheduled( self::POST_SETUP_SUCCESS_ACTION ) ) {
 				wp_schedule_single_event( time() + DAY_IN_SECONDS, self::POST_SETUP_SUCCESS_ACTION );
 			}
```
Are we sure it's best to `return` here when an exception is raised? I'm wondering if there's any need to go into the `if ( $verification_complete )` branches.

I see that the previous implementation _effectively_ has a return statement, but at that point this check had already been done. Maybe we need to make sure that part of the code is run here as well?

ricardo replied:
I think the `return` can be removed. Will do it 🙂.
I didn't know the reason for wrapping those in try/catch blocks, so I assumed if the `WC_Data_Store` or `WC_Admin_Notes` can fail, it will probably fail after the catch too in `should_show_marketing_note` or `create_failure_note`.

reykjalin replied:
Yeah, I'm not sure about that either. Let's assume for now that we might want it to run through that if statement; just like it did before 🙂

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -33,15 +40,17 @@ class WC_Stripe_Apple_Pay_Registration {
-		add_action( 'admin_notices', array( $this, 'admin_notices' ) );
 
 		$this->stripe_settings         = get_option( 'woocommerce_stripe_settings', array() );
+		$this->domain_name             = isset( $_SERVER['HTTP_HOST'] ) ? $_SERVER['HTTP_HOST'] : str_replace( [ 'https://', 'http://' ], '', get_site_url() ); // @codingStandardsIgnoreLine
```
I think this is a good case for the [null coalescing operator](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce) 🙂 

```suggestion
		$this->domain_name             = $_SERVER['HTTP_HOST'] ?? str_replace( [ 'https://', 'http://' ], '', get_site_url() ); // @codingStandardsIgnoreLine
```

ricardo replied:
We could use it, but apparently this operator is a relatively new thing and was introduced in PHP 7.0. We still need to support PHP 5.2, so I believe we shouldn't use it.

https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op

reykjalin replied:
Ah, of course. I'm used to being able to use these newer features in WooCommerce Payments. Just ignore this! 😛

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -90,6 +99,62 @@ private function get_secret_key() {
+			return;
+		}
+
+		$error = null;
+		if ( ! file_exists( $path . '/' . $dir ) ) {
+			if ( ! @mkdir( $path . '/' . $dir, 0755 ) ) { // @codingStandardsIgnoreLine
+				$error = __( 'Unable to create domain association folder to domain root.', 'woocommerce-payments' );
+			}
+		}
+
+		if ( ! @copy( WC_STRIPE_PLUGIN_PATH . '/' . $file, $fullpath ) ) { // @codingStandardsIgnoreLine
+			$error = __( 'Unable to copy domain association file to domain root.', 'woocommerce-payments' );
+		}
```
I think this might be better as a function? The function can still return `$error` or `null`. Or maybe it'd be better to go with our recent-ish team policy of throwing an exception? I'd love to get your thoughts on that.

Point being that we could do something like

```php
$error = $this->copy_and_overwrite_domain_association_file()
if ( isset( $error ) ) {
// ...
```
or
```php
try {
  $this->copy_and_overwrite_domain_association_file();
} catch ( ApplePayDomainUpdateError $error ) {
  log( $error->message );
}
```

I think that'd make the flow here a little bit cleaner. What do you think?

ricardo replied:
This whole `update_domain_association_file` function was restored from a previous implementation and just adapted here as a fallback verification method.

In order to separate these functions I need to do something about the shared variables. I'll suggest something and then you can review.

ricardo replied:
@reykjalin, I agree this function is not so easy to read. I started refactoring it here, but it started looking worse, because we have to pass those shared variables or re-declare them ($dir, $file, $path, $fullpath) and the first verification at the top...

```php
$existing_contents = @file_get_contents( $fullpath ); // @codingStandardsIgnoreLine
$new_contents      = @file_get_contents( WC_STRIPE_PLUGIN_PATH . '/' . $this->domain_association_ile ); // @codingStandardsIgnoreLine
if ( $existing_contents === $new_contents ) {
    return;
}
```

... ideally would be split too, so we'd have 4 different functions in total, thus increasing the overall complexity of this class.

After some thought, I'd suggest keeping it as is, unless you feel strongly about it and/or have a suggestion on how to make it cleaner with the shared variables in mind  🙂

ricardo replied:
FYI the original function can be found [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1357/files#diff-b3ce8b937d7eea48160e4d6e7bd28e652b0b4834af86ac2b60bdb748aac0eedfL181). Now we're adding it again as a fallback because the rewrite rule isn't always reliable.

reykjalin replied:
I'm personally fine with creating a private helper function like `get_well_known_directory_path()`:

```php
function get_well_known_directory_path() {
	return untrailingslashit( ABSPATH ) . '/.well-known';
}
```

We can then use that helper function in `update_domain_association_file()`:

```php
function copy_and_overwrite_domain_association_file() {
	$well_known_dir = $this->get_well_known_directory_path();

	// This might even be added to a `get_local_domain_association_path()` helper function.
	$file = 'apple-developer-merchantid-domain-association';
	$fullpath = $well_known_dir . '/' . $file;

	if ( ! file_exists( $well_known_dir ) ) {
		if ( ! @mkdir( $well_known_dir, 0755 ) ) { // @codingStandardsIgnoreLine
			return __( 'Unable to create domain association folder to domain root.', 'woocommerce-payments' );
		}
	}

	if ( ! @copy( WC_STRIPE_PLUGIN_PATH . '/' . $file, $fullpath ) ) { // @codingStandardsIgnoreLine
		return __( 'Unable to copy domain association file to domain root.', 'woocommerce-payments' );
	}
}
```

Sure, it adds a bit more noise in the form of functions, but it will make the actual business logic easier to read, I think.

My overall worry here is that the work being done in the current function is difficult to parse; I want to make that as easy to follow as possible. If that means hiding the complexities in private functions that add some noise then so be it 🙂

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -90,6 +99,62 @@ private function get_secret_key() {
+		}
+
+		if ( isset( $error ) ) {
+			$url            = get_site_url() . '/' . $dir . '/' . $file;
+			$response       = @wp_remote_get( $url ); // @codingStandardsIgnoreLine
+			$already_hosted = @wp_remote_retrieve_body( $response ) === $new_contents; // @codingStandardsIgnoreLine
```
It'd also be nice to see this extracted into a function like `verify_hosted_domain_association_file_is_up_to_date()` or something along those lines.

We could then make the flow here a bit cleaner:

```php
if ( isset( $error ) && ! $this->verify_hosted_domain_association_file_is_up_to_date() ) {
  log( $error );
}
```

Or if we decide to go with exception handling above (see other comment):

```php
try {
  $this->copy_and_overwrite_domain_association_file();
} catch ( ApplePayDomainUpdateError $error ) {
  if ( ! $this->verify_hosted_domain_association_file_is_up_to_date() ) {
    log( "Some different error and " . $error->message );
    return;
  }
  log( $error->message );
}
```

ricardo replied:
Please see my [previous comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1466#discussion_r581153402).

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -90,6 +99,62 @@ private function get_secret_key() {
+			}
+		}
+
+		WC_Stripe_Logger::log( __( 'Domain association file updated.', 'woocommerce-payments' ) );
```
Is it intentional to log this even if an error happens?

ricardo replied:
Not really. Good catch. 👍

reykjalin commented on tests/phpunit/test-wc-stripe-apple-pay-registration.php:
```diff
@@ -0,0 +1,77 @@
+		$fullpath = $path . '/' . $dir . '/' . $this->file_name;
+
+		$this->wc_apple_pay_registration->update_domain_association_file();
+		$expected_file_contents = file_get_contents( $fullpath ); // @codingStandardsIgnoreLine
+
+		$this->assertEquals( $expected_file_contents, $this->file_contents );
```
Just for clarity; I think the variable names here should be reversed. The variable called `$expected_file_contents` contains the new updated value after running `update_domain_association_file()`, whereas `$this->file_contents` is what was in there before (i.e. the expected value) 🙂 

So maybe rename `$expected_file_contents` to `$new_file_contents` or `$updated_file_contents`? 
`$this->file_contents` might also be better served as being called `$this->initial_file_contents` or something along those lines.

ricardo replied:
Agreed. Will update.

reykjalin reviewed (changes requested):
Awesome work, especially for adding the tests! 👏 🎉 

I left a couple of comments where I think we can improve the code in terms of readability. I still haven't tested this, since I need to go through the whole new site dance to get that working, so I'll add a different comment/review once I've done that 🙂 

For now, just discussions around code readability ☺️

reykjalin commented:
Everything works as advertised during testing. Fantastic work @ricardo ! 👏 🥇 

Let's just address the outstanding discussions and get this merged 🎉

ricardo committed:
Don't log success on domain association file update if there was an error

ricardo committed:
Remove return from catch block on inbox notice

ricardo committed:
Rename variables in tests

ricardo committed:
Add changelog entries

ricardo commented:
@reykjalin

- I added some comments;
- Implemented suggested changes and rebased;
- Added 2 changelog entries, please let me know how that sounds.

reykjalin reviewed (commented):
Changelog looks good 👍

ricardo committed:
Refactor update_domain_association_file

ricardo committed:
Merge branch 'trunk' into fix/1442-apple-pay-registration-fallback

ricardo commented:
@reykjalin I've refactored that part of the code and tested everything again, let me know what you think.

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -33,15 +44,17 @@ class WC_Stripe_Apple_Pay_Registration {
 
 	public function __construct() {
 		add_action( 'init', array( $this, 'add_domain_association_rewrite_rule' ) );
+		add_action( 'admin_init', [ $this, 'verify_domain_on_domain_name_change' ] );
```
Sorry to just comment on this now; Let's just the long array syntax `array()` instead of the short `[]` for uniformity; we're using the long syntax (seemingly) everywhere.

reykjalin commented on includes/class-wc-stripe-apple-pay-registration.php:
```diff
@@ -33,15 +44,17 @@ class WC_Stripe_Apple_Pay_Registration {
-		add_action( 'admin_notices', array( $this, 'admin_notices' ) );
 
 		$this->stripe_settings         = get_option( 'woocommerce_stripe_settings', array() );
+		$this->domain_name             = isset( $_SERVER['HTTP_HOST'] ) ? $_SERVER['HTTP_HOST'] : str_replace( [ 'https://', 'http://' ], '', get_site_url() ); // @codingStandardsIgnoreLine
```
Same as above; let's use the long-form array syntax.

reykjalin reviewed (changes requested):
Just one last little thing and then we can ship! 🎉

ricardo committed:
Fixing array syntax

ricardo committed:
Merge branch 'fix/1442-apple-pay-registration-fallback' of https://github.com/woocommerce/woocommerce-gateway-stripe into fix/1442-apple-pay-registration-fallback

reykjalin reviewed (approved):
Looks good! Let's get this shipped! 🚀
